File: Course_2-Data_Structures/Week-1/Excercise_Challenges/1_brackets_in_code/check_brackets.py
def matching(left, right):
    return (left + right) in ["()", "[]", "{}"]

def is_balanced(text):
    stack = []
    for i, c in enumerate(text):
        if c in ['(', '{', '[']:
            stack.append([c, i])
        else:
            if c in [')', '}', ']']:
                if len(stack) == 0:
                    return i + 1
                top = stack.pop()[0]
                if matching(top, c) == False:
                    return i + 1
                
    if len(stack):
        return stack.pop()[1] + 1
    else:
        return 0


if __name__ == "__main__":
    text = input()
    result = is_balanced(text)

    if result:
        print(result)
    else:
        print("Success")

File: Course_2-Data_Structures/Week-1/Excercise_Challenges/1_brackets_in_code/check_brackets.cpp
#include <iostream>
#include <stack>
#include <string>
#include <utility> 
#define MP make_pair
#define F first
#define S second
using namespace std;

bool cmp(char a, char b)
{
    if(a == '(' && b == ')')
        return true;
    else if(a == '{' && b == '}')
        return true;
    else if(a == '[' && b == ']')
        return true;
    else
        return false;
}

int is_balanced(string text)
{
    stack<pair<char, int> > s;
    int n = text.length();

    for(int i = 0; i < n; ++i){
        if(text[i] == '(' || text[i] == '{' || text[i] == '[')
            s.push(MP(text[i], i));
        else{
            if(text[i] == ')' || text[i] == '}' || text[i] == ']'){
                if(s.empty())
                    return i + 1;
                
                char top = s.top().F;
                s.pop();
                if(cmp(top, text[i]) == false)
                    return i + 1;
            }
        }
    }

    return (s.empty()? -1 : s.top().S + 1);
}

int main(void)
{
    string text;
    getline(cin, text);

    int result = is_balanced(text);
    if(result == -1)
        cout << "Success" << "\n";
    else
        cout << result << "\n";

    return 0;
}
